src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Pool.cs

```diff
@@ -5,63 +5,63 @@
 using System.Diagnostics;
 using System.Threading;
 
-namespace Microsoft.Data.SqlClient.SNI
+namespace Microsoft.Data.SqlClient
 {
     // this is a very simple threadsafe pool derived from the aspnet/extensions default pool implementation
     // https://github.com/dotnet/extensions/blob/release/3.1/src/ObjectPool/src/DefaultObjectPool.cs
-    internal sealed class SNIPacketPool
+    internal sealed class Pool<T> where T : class
```

Comment:
It is purposefully generic. It's a pool of T whatever T happens to be. It's a location to store relatively frequently used or expensive objects. It already works for packets but it can be used for all sorts of things like SqlEnvChange which we'll want frequently and are wasteful to allocate and then GC almost immediately when we'll need another one soon.

T is constrainted to class because it makes no sense to pool a valuetype, they're copied, and because if you don't constrain it the compiler won't allow you to properly do null checks because they wouldn't make sense if T could be either a class or a value. Yes you could declare T as an interface of abstract class if you wanted to but clearly you'd only be able to cache concrete instances.

I'm taking the packet pool and making the pool aspect of it reusable. It's supposed to be generic. The naming of the internal variables comes from the original aspnet version, see the comment at the top of the file.

Reply:
I'd suggest calling it "SqlClientObjectPool" so it doesn't mix up with non-driver types.

Reply:
It's internal and we don't use any other pool implementations nor would we be able to pull in any implementations without taking a dependency on an upstack library like runtime extensions at this point, that seems unlikely. Adding SqlClient to the name seems pointless when it's already part of the namespace.

Reply:
There's a reason why libraries prefix their own classes, and there're many classes in our driver prefixed accordingly. It's main goal is to add clarity to reader/developer that this is driver's own implementation. `SqlObjectPool` is also an option, but just leaving it as "Pool" doesn't go well with product development practices. Only Util classes are exempted from such naming conventions.

Also you mentioned above, 

> It already works for packets but it can be used for all sorts of things like SqlEnvChange which we'll want frequently and are wasteful to allocate and then GC almost immediately when we'll need another one soon.

If we do want to use it for other cases, rather make is follow good practices so it can be considered as a new generic pool implementation. I understand it's generic code, but it's not generic enough to be already available from Dotnet SDK. That's why if we implement, we own it and name it accordingly, internal/public does not matter.

Reply:
Renamed to `SqlObjectPool`